src/test/com/Elessar/app/client/MyClientIntegTest.java

```diff
@@ -0,0 +1,233 @@
+        final MetricManager clientMetricManager = new MetricManager("ClientMetric", 1000000);
+
+        // Setup server
+        mongoDB = MongoClients.create("mongodb://localhost:27017").getDatabase("test");
+        mongoDB.getCollection(MyDatabase.USERS).createIndex(Indexes.text(User.NAME), new IndexOptions().unique(true));
+
+        final MyDatabase db = new MongoDB(mongoDB, serverMetricManager);
+        final MyServer server = new MyServer("localhost", serverPort, db, serverMetricManager);
+
+        // Setup client A
+        final BlockingQueue<String> msgQueueA = new LinkedBlockingQueue<>();
+        clientA = new MyClient(serverURL, clientMetricManager);
+        final MyClientServer clientA_Server = new MyClientServer("localhost", clientA_Port, msgQueueA, clientMetricManager);
+
+        // Setup client B
+        final BlockingQueue<String> msgQueueB = new LinkedBlockingQueue<>();
+        clientB = new MyClient(serverURL, clientMetricManager);
+        final MyClientServer clientB_Server = new MyClientServer("localhost", clientB_Port, msgQueueB, clientMetricManager);
+
+        server.run();
+        clientA_Server.run();
+        clientB_Server.run();
+    }
+
+
+    @Test
+    public void test() {
+        registerTest();
+
+        logOnTest();
+
+        logOffTest();
+
+        p2pMsgTest();
+    }
+
+    private void registerTest() {
+        try {
+            try (MongoCursor<Document> cursor = mongoDB.getCollection(MyDatabase.USERS).find(or(eq(User.NAME, "Shuai Hu"),
+                    eq(User.NAME, "Zhi Xu"))).iterator()) {
+                assertEquals(false, cursor.hasNext());
+            }
+
+            RegistrationResponse registerResp1 = clientA.register("Shuai Hu", "4238902", "shuaih@yahoo.com", "62357492");
+            assertEquals(true, registerResp1.getSuccess());
+
+            RegistrationResponse registerResp2 = clientA.register("Shuai Hu", "4238902", "shuaih@yahoo.com", "62357492");
+            assertEquals(false, registerResp2.getSuccess());
+
+            RegistrationResponse registerResp3 = clientA.register("Zhi Xu", "213252", "zhix@126.com", "62258237");
+            assertEquals(true, registerResp3.getSuccess());
+
+            try (MongoCursor<Document> cursor = mongoDB.getCollection(MyDatabase.USERS).find(eq(User.NAME, "Shuai Hu")).iterator()) {
+                assertEquals(true, cursor.hasNext());
+            }
+
+            try (MongoCursor<Document> cursor = mongoDB.getCollection(MyDatabase.USERS).find(eq(User.NAME, "Zhi Xu")).iterator()) {
+                assertEquals(true, cursor.hasNext());
+            }
+
+        } catch (Exception e) {
+            System.out.println(e.getMessage());
+            assert(false);
+        }
+    }
+
+
+    private void logOnTest() {
+        try {
+            try (MongoCursor<Document> cursor = mongoDB.getCollection(MyDatabase.USERS).find(eq(User.NAME, "Shuai Hu")).iterator()) {
+                if (cursor.hasNext()) {
```

Comment:
Probably better to cleanup the collection and do the setup again here.
We can have 50 private methods testing different things and someone might add a new test method in the middle and accidentally breaks all tests after it.

Here if registrationTest() fails and Users table doesn't have this user we won't even notice something is wrong as we are only doing the assertion inside the if.

Same for the other tests